netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java

```diff
@@ -0,0 +1,305 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.transport.netty;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.netty.handler.codec.http2.Http2Error.PROTOCOL_ERROR;
+import static io.netty.handler.codec.http2.Http2Exception.connectionError;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http2.DefaultHttp2ConnectionEncoder;
+import io.netty.handler.codec.http2.Http2CodecUtil;
+import io.netty.handler.codec.http2.Http2Connection;
+import io.netty.handler.codec.http2.Http2ConnectionAdapter;
+import io.netty.handler.codec.http2.Http2ConnectionEncoder;
+import io.netty.handler.codec.http2.Http2Exception;
+import io.netty.handler.codec.http2.Http2Flags;
+import io.netty.handler.codec.http2.Http2FrameWriter;
+import io.netty.handler.codec.http2.Http2Headers;
+import io.netty.handler.codec.http2.Http2LifecycleManager;
+import io.netty.handler.codec.http2.Http2RemoteFlowController;
+import io.netty.handler.codec.http2.Http2Settings;
+import io.netty.handler.codec.http2.Http2Stream;
+import io.netty.util.ReferenceCountUtil;
+
+import java.util.ArrayDeque;
+import java.util.Map;
+import java.util.Queue;
+import java.util.TreeMap;
+
+/**
+ * <p>
+ * Implementation of a {@link Http2ConnectionEncoder} that dispatches all method call to
+ * another {@link Http2ConnectionEncoder}, except for when the maximum number of
+ * concurrent streams limit is reached.
+ * </p>
+ * <p>
+ * When this limit is hit, instead of rejecting any new streams this implementation
+ * buffers newly created streams and their corresponding frames. Once an active stream
+ * gets closed or the maximum number of concurrent streams is increased, this encoder will
+ * automatically try to empty its buffer and create as many new streams as possible.
+ * </p>
+ * <p>
+ * This implementation makes the buffering mostly transparent and can be used as a drop
+ * in replacement for {@link DefaultHttp2ConnectionEncoder}.
+ * </p>
+ */
+class BufferingHttp2ConnectionEncoder extends Http2ConnectionEncoderDecorator {
+
+  // Buffer for any streams and corresponding frames that could not be created
+  // due to the maximum concurrent stream limit being hit.
+  private final TreeMap<Integer, PendingStream> pendingStreams =
```

Comment:
@nmittler whatever data structure we use, we need to be able to fetch the next stream (stream with the lowest stream id). A hash map can't do this. So the choice was to either use a queue and a hash map or to use a tree map. I went with the tree map. WDYT?